### PR TITLE
start free associative algebra

### PR DESCRIPTION
**Pull request by tthsqe12**
This is a good start: basic interface and arithmetic up to divexact.

### CODE REVIEW

File: src/generic/FreeAssAlgebra.jl [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+   return length(a.S)
+end
+
+function length(a::FreeAssAlgElem)
+   return a.length
+end
+
+function check_parent(a::FreeAssAlgElem{T}, b::FreeAssAlgElem{T}, throw::Bool = true) where T <: RingElement
+   b = parent(a) != parent(b)
+   b & throw && error("Incompatible polynomial rings in polynomial operation")
```
Comment (wbhart) [inline src/generic/FreeAssAlgebra.jl:45]: Probably shouldn't refer to it as a polynomial ring.

**Comment by wbhart** [comment]
Apart from a few unclear things, this is looking good so far, modulo the lack of tests and documentation of course.

**Comment by tthsqe12** [comment]
@thofma what about a second parameter U for both the parent and elems with the exps::Vector{Vector{U}}, and a half type unstable constructor? U = UInt8 when nvars<256 for example.

File: src/generic/FreeAssAlgebra.jl (hunk 1 above)
Comment (tthsqe12) [reply src/generic/FreeAssAlgebra.jl:45]: sure.

**Comment by thofma** [comment]
OK. If you only want to support `UInt8` or `fmpz` you might also consider just putting two entries in the struct, with one being `undef`.

**Comment by tthsqe12** [comment]
I don't like `undef`. And, plain `Int` will always suffice. So, it its really just a choice between `UInt8` and `Int`

File: src/generic/FreeAssAlgebra.jl [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+# return c*w*a*wp
+function mul_term(c::T, w::Vector{Int}, a::FreeAssAlgElem{T}, wp::Vector{Int}) where T
+   zcoeffs = isone(c) ? T[a.coeffs[i] for i in 1:a.length] :
+                        T[c*a.coeffs[i] for i in 1:a.length]
+   zexps = Vector{Int}[vcat(w, a.exps[i], wp) for i in 1:a.length]
+   return FreeAssAlgElem{T}(parent(a), zcoeffs, zexps, a.length)
+end
+
+
+# return (true, l, r) with a = l*b*r and length(l) minimal
```
[the remark this answers is not in the document]
Comment (tthsqe12) [reply src/generic/FreeAssAlgebra.jl:429]: left most division of monomials. I use "word" for the representation of monmials as Vector{Int}

**Comment by tthsqe12** [comment]
I think this is good now

**Comment by thofma** [comment]
What is up with `leading_*`? Is there an ordering? It is not mentioned anywhere.

**Comment by tthsqe12** [comment]
> What is up with leading_*? Is there an ordering? It is not mentioned anywhere.

I could mention that the terms are stored in a definite order. I would have to figure out what it is, and then I could add it to the documentation.

**Comment by thofma** [comment]
Sounds like an implementation detail. Should we expose this?

**Review by thofma — APPROVED** [review]

**Pull request merged**